### ISSUE TITLE
BatchNorm has a biased estimation

### ISSUE THREAD

**Issue by YUE-FAN** [kekmodel/FixMatch-pytorch#7]
Hi :)

In the function train() in train.py, I think batch normalization layers will have a biased estimation if you feed the concatenated inputs to the model. It should rather be like:

```
inputs = torch.cat((inputs_x, inputs_u_s)).to(args.device)
targets_x = targets_x.to(args.device)
logits = model(inputs)
logits_x , logits_u_s = logits.chunk(2)
model.eval()
logits_u_w= model(inputs_u_w)
model.train()        
del logits
```

Besides, it is mentioned in the paper that they also applied the unlabeled loss for labeled data

> In practice, we include all labeled examples as part of unlabeled data
without using their labels when constructing U.

**Comment by kekmodel** [comment]
[](https://www.dropbox.com/s/7f8amr12cy8p5za/Screenshot%20from%202020-04-17%2010-26-08.png?dl=0)

**Comment by YUE-FAN** [comment]
Thanks for pointing me to the original code. Yours is indeed correct. We do not need `model.eval()` for batch normalization.

However, when I checked the original code, I found they include all labeled examples as part of unlabeled data, and they also take a part of the unlabeled data as val data.
```
CUDA_VISIBLE_DEVICES= scripts/create_unlabeled.py $ML_DATA/SSL2/cifar10 $ML_DATA/cifar10-train.tfrecord
CUDA_VISIBLE_DEVICES= scripts/create_split.py --seed=$seed --size=$size $ML_DATA/SSL2/cifar10 $ML_DATA/cifar10-train.tfrecord
```
These details seem to be missing from this implementation. Please correct me if I am wrong.
Thanks in advance.

**Comment by jamesbsilva** [comment]
> [FixMatch Algotirhm pseudo code](https://www.dropbox.com/s/7f8amr12cy8p5za/Screenshot%20from%202020-04-17%2010-26-08.png?dl=0)
> [Official Code](https://github.com/google-research/fixmatch/blob/08d9b83d7cc87e853e6afc5a86b12aacff56cdea/fixmatch.py#L96)

I was wondering what the impact is from line 7 of the pseudo code. 

This seems to happen in the original code as well but it seems that both implementations calculate the average for the unsupervised loss that treats the masked parts as zero rather than excluding them in the average.

Have you looked into the impact of this?

averaging a mask of `[1,0,1]`

so that the average of masking `[1,2,3]` is calculated as the average of `[1,0,3]` instead of `[1,3]` does the small difference in scale matter?

**Comment by kekmodel** [comment]
> > [FixMatch Algotirhm pseudo code](https://www.dropbox.com/s/7f8amr12cy8p5za/Screenshot%20from%202020-04-17%2010-26-08.png?dl=0)
> > [Official Code](https://github.com/google-research/fixmatch/blob/08d9b83d7cc87e853e6afc5a86b12aacff56cdea/fixmatch.py#L96)
> 
> I was wondering what the impact is from line 7 of the pseudo code.
> 
> This seems to happen in the original code as well but it seems that both implementations calculate the average for the unsupervised loss that treats the masked parts as zero rather than excluding them in the average.
> 
> Have you looked into the impact of this?
> 
> averaging a mask of `[1,0,1]`
> 
> so that the average of masking `[1,2,3]` is calculated as the average of `[1,0,3]` instead of `[1,3]` does the small difference in scale matter?

It makes sense. You can test it by modifying the loss.
